cupy/random/cupy_distributions.cu

```diff
@@ -255,14 +256,50 @@ struct standard_normal_float_functor {
     }
 };
 
+// There are several errors when trying to do this a full template
+struct standard_gamma_functor {
+    template<typename... Args>
+    __device__ double operator () (Args&&... args) {
+        return rk_standard_gamma(args...);
+    }
+};
+
+// The following templates are used to unwrap arrays into an elementwise
+// approach
+// When a pointer is present in the variadic Args, it will be replaced by
+// the value of pointe[thread_id]
+template<bool is_pointer, typename T>
+struct pointer_fnct {};
+
+template<typename T>
+struct pointer_fnct<true, T> {
+    static __device__ double get(int64_t *value, int id) {
+        void *ptr = reinterpret_cast<void*>(value[0]);
+        int ndim = value[1];
+        ptrdiff_t offset = 0;
+        for (int dim = ndim; --dim >= 0; ) {
+            offset += value[ndim + dim + 2] * (id % value[dim + 2]);
+            id /= value[dim + 2];
+        }
+        return *reinterpret_cast<double*>(ptr + offset);
+    }
+};
+
+template<typename T>
+struct pointer_fnct<false, T> {
+    static __device__ T get(T value, int id) {
+        return value;
+    }
+};
+
 template<typename F, typename T, typename R, typename... Args>
 __global__ void execute_dist(intptr_t state, intptr_t out, ssize_t size, Args... args) {
     int id = threadIdx.x + blockIdx.x * blockDim.x;
     R* out_ptr = reinterpret_cast<R*>(out);
     if (id < size) {
         T random(id, state);
         F func;
-        out_ptr[id] = func(&random, std::forward<Args>(args)...);
+        out_ptr[id] = func(&random, (pointer_fnct<std::is_pointer<Args>::value, Args>::get(args, id))...);
```

Comment:
Silly question: If we check `Args` with `std::is_pointer`, doesn't it mean `Args` is unpack with only 1 type? Why don't we just make it, say,
```c++
template<typename F, typename T, typename R, typename X>
```
and avoid packing/unpacking?

Reply:
The function that applies this is also a variadic template, so what we do is inspect all args and when we have a pointer, it means is a cupy array, so we do the elementwise access in compile time :)

Reply:
Ah OK. I didn't understand why the template type `Args` has to be variadic, but now I see `execute_dist` is called a few lines below with forwarded variadic arguments. Thanks.